Add `varLib.hvar` module

behdad opened the pull request:
Fixes https://github.com/fonttools/fonttools/issues/3778

Mildly tested. In particular, on RobotoFlex VF, it produces smaller HVAR table. I have not investigated. The code is heavily lifted from `varLib.__init__`.

anthrotype commented on Lib/fontTools/varLib/hvar.py:
```diff
@@ -0,0 +1,167 @@
+        storeBuilder.setSupports(supports)
+        advMapping[glyphName] = storeBuilder.storeDeltas(deltas, round=noRound)
+
+    singleModel = all(advMapping[g] >> 16 == 0 for g in glyphOrder)
```
shouldn't this check that all the supports are the same (like we do in the varLib _get_advance_metrics)?

behdad replied:
That's one way. My reasoning here was that if the OnlineVarStoreBuilder put them all in the same group (ie. all major's match), then we can do this. I'll try to revert to other logic.

anthrotype commented on Lib/fontTools/varLib/hvar.py:
```diff
@@ -0,0 +1,167 @@
+            phantoms[1] = phantoms[1][phantomIndex] if phantoms[1] is not None else 0
+            deltas.append(phantoms[1] - phantoms[0])
+
+        allDeltas[glyphName] = deltas[1:]
```
why skip the first one? an inline comment might help the reader

behdad replied:
First "delta" is the actual advance itself.

anthrotype commented:
I wish we could reuse more of the code that's already in varLib.__init__ (e.g. `_get_advance_metrics`) instead of maintaining two very similar copies

behdad commented on Lib/fontTools/varLib/hvar.py:
```diff
@@ -0,0 +1,167 @@
+    directStore = None
+    if singleModel:
+        # Build direct mapping
+        supports = supports[1:]
```
Ah. I think I messed up both of these. I'll fix. I used to have {} as first item, but I think I removed it.

behdad commented:
> I wish we could reuse more of the code that's already in `varLib.__init__` (e.g. `_get_advance_metrics`) instead of maintaining two very similar copies

Me too. I'll work on it.

behdad commented:
I addressed reviews.

One thing occurs to me that in a post-64k world, we should be more careful building the direct mapping since not all will fit in a VarData.

I'll see about more code sharing. And needs some kinds of tests.

behdad commented:
I haven't found a font yet that this fully roundtrips to the same XML...

behdad commented:
> I haven't found a font yet that this fully roundtrips to the same XML...

We changed the sort algorithm a bit. Maybe it's that.

behdad commented:
> > I haven't found a font yet that this fully roundtrips to the same XML...
> 
> We changed the sort algorithm a bit. Maybe it's that.

Oh it's probably because we don't go through the VariationModel, just directly to deltas. Anyway, I've convinced myself that the code is correct. Now onto refactoring.

behdad commented:
VORG is not done in this module. VVAR is done but untested.

anthrotype commented on Lib/fontTools/varLib/__init__.py:
```diff
@@ -657,10 +600,71 @@ def _get_advance_metrics(
+        )
+
+    if vOrigMetricses:
+        singleModel = False
```
local variable no longer used within this scope

anthrotype commented:
> I haven't found a font yet that this fully roundtrips to the same XML...

maybe because we end up ordering the VarRegions differently in the VarRegionList?

> And needs some kinds of tests.

yes, ideally

behdad commented:
> > I haven't found a font yet that this fully roundtrips to the same XML...
> 
> maybe because we end up ordering the VarRegions differently in the VarRegionList?

I built a fresh NotoSansArabic, and HVAR roundtrips perfectly.
